src/config_format/flb_cf_yaml.c

```diff
@@ -1338,114 +1408,252 @@ static char *get_real_path(char *file, char *path, size_t size)
     }
+    state->section = last->section;
+    state->keyvals = last->keyvals;
+    state->cf_section = last->cf_section;
+    state->cf_group = last->cf_group;
+    state->values = last->values;
+    state->file = last->file;
+    state->state = state_num;
+    state->level = last->level + 1;
+    state->key = last->key;
+
+    cfl_list_add(&state->_head, &ctx->states);
+    return state;
+}
 
-    if (s->file) {
-        flb_sds_destroy(s->file);
+static struct parser_state *state_push_section(struct local_ctx *ctx,
+                                               enum state state_num,
+                                               enum section section)
+{
+    struct parser_state *state;
+
+    state = state_push(ctx, state_num);
+    if (state == NULL) {
+        return NULL;
     }
+    state->section = section;
 
-    if (s->root_path) {
-        flb_sds_destroy(s->root_path);
+    return state;
+}
+
+static struct parser_state *state_push_key(struct local_ctx *ctx,
+                                           enum state state_num,
+                                           const char *key)
+{
+    struct parser_state *state;
+    flb_sds_t skey;
+
+    skey = flb_sds_create(key);
+    if (skey == NULL) {
+        return NULL;
     }
-    flb_free(s);
+    state = state_push(ctx, state_num);
+    if (state == NULL) {
+        flb_sds_destroy(skey);
+        return NULL;
+    }
+    state->key = skey;
+    state->allocation_flags |= HAS_KEY;
+    return state;
 }
 
-static struct parser_state *state_create(char *caller_file, char *file)
+static struct parser_state *state_push_withvals(struct local_ctx *ctx,
+                                                struct parser_state *parent,
+                                                enum state state_num)
 {
-    int ret;
-    char *p;
-    char file_path[PATH_MAX + 1] = {0};
-    char caller_path[PATH_MAX + 1] = {0};
-    struct parser_state *s;
-    struct stat st;
+    struct parser_state *state;
+    struct cfl_kvlist *kvlist;
 
-    if (!file) {
+    kvlist = cfl_kvlist_create();
+    if (kvlist == NULL) {
+        return NULL;
+    }
+    state = state_push(ctx, state_num);
+    if (state == NULL) {
+        cfl_kvlist_destroy(kvlist);
         return NULL;
     }
+    state->keyvals = kvlist;
+    state->allocation_flags |= HAS_KEYVALS;
 
-    /* allocate context */
-    s = flb_calloc(1, sizeof(struct parser_state));
-    if (!s) {
-        flb_errno();
+    return state;
+}
+
+static struct parser_state *state_push_witharr(struct local_ctx *ctx,
+                                               struct parser_state *parent,
+                                               enum state state_num)
+{
+    struct parser_state *state;
+
+    parent->values = cfl_array_create(4);
+    if (parent->values == NULL) {
+        flb_error("no value");
+        return YAML_FAILURE;
+    }
+    cfl_array_resizable(parent->values, CFL_TRUE);
+
+    state = state_push(ctx, state_num);
```

Comment:
Don't you need to dispose of `parent->values` if this call fails?

Reply:
No, the values should get released later on when the state returns when the parent state is freed.